### PR TITLE
[Tune] [PBT] [Doc] Fix and clean up PBT examples

### LINKED ISSUE #22733: [Bug] Train PBT example is not using checkpointing
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Train

### What happened + What you expected to happen

Train PBT example is not using checkpointing, which is necessary for PBT to work as intended.

### Versions / Dependencies

Ray 1.11

### Reproduction script

N/A

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

Many examples in the docs related to PBT either did not have checkpointing or were checkpointing/loading incorrectly. The examples updated are:
- PBT User Guide (originally `tune-advanced-tutorial`) (converted this to a [runnable notebook](https://colab.research.google.com/github/justinvyu/ray/blob/justinvyu/train-pbt-test-checkpoint/doc/source/tune/examples/pbt_guide.ipynb) with the fixes)
- `pbt_function`
- `pbt_example`
- `pbt_memnn_example`
- `pbt_tune_cifar10_with_keras`
- `pb2_example`
    - This example uses the same training function as `pbt_function`, so some of the parameters needed to be updated.
- `tune_cifar_torch_pbt_example`
    - This one affects the long-running `pytorch_pbt_failure` release test (it uses the train function defined in this example).

This is **especially the case with examples that use the Function Trainable API with PBT, since we require the user to checkpoint themselves** via `session.report`. This makes it a bit harder to align checkpoints with PBT perturbations. The main issue is that it's difficult for the user to keep track of the iteration number themselves (which is what they need to do if they want to copy the `checkpoint_frequency` functionality that is available when using the class Trainable API).

Consider the following cases:

1. The starting step needs to be set to 1. Otherwise, checkpointing and perturbation will be out of sync:

```python
step = 0

# Checkpoint every `checkpoint_interval` steps
if step % checkpoint_interval == 0:
    # NOTE: Since we initialized `step=0` above, our checkpointing and perturbing
    # are out of sync by 1 step.
    # Ex: if `checkpoint_interval` = `perturbation_interval` = 3
    # step:                0 (checkpoint)  1     2            3 (checkpoint)
    # training_iteration:  1               2     3 (perturb)  4
    checkpoint = Checkpoint.from_dict({"acc": accuracy, "step": step})
session.report(..., checkpoint=checkpoint)
step += 1
```

vs.

```python
step = 1

# Checkpoint every `checkpoint_interval` steps
if step % checkpoint_interval == 0:
    # Fixed if we initialize step = 1
    # Ex: if `checkpoint_interval` = `perturbation_interval` = 3
    # step:                1          2     3 (checkpoint)     4
    # training_iteration:  1          2     3 (perturb)        4
    checkpoint = Checkpoint.from_dict({"acc": accuracy, "step": step})
session.report(..., checkpoint=checkpoint)
step += 1
```

2. The user can easily start from the wrong step upon restore if they don't increment the checkpointed step by 1.

```python
if session.get_checkpoint():
    state = session.get_checkpoint().to_dict()
    accuracy = state["acc"]
    last_step = state["step"]
    # Current step should be 1 more than the last checkpoint step
    # If we did `step = last_step` instead, we might repeat the step and end up
    # checkpointing more than we want to.
    # Ex: last_step = 4, step = 4 --> if `checkpoint_interval = 4`,
    # then we would checkpoint again, even though we just restored.
    # Should be last_step = 4, step = 5 --> next checkpoint will be step = 8
    step = last_step + 1
```

### Open questions:
- Would this issue of requiring the user to manually keep another `step` counter be solved if we introduced a `session.get_training_iteration()` API?
  - One problem with this is that the user needs to create the checkpoint before calling `session.report`, and `session.report` is what increments the training iteration.

## Related issue number

Closes https://github.com/ray-project/ray/issues/22733

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/examples/tune_cifar_torch_pbt_example.py [hunk 1]
```diff
@@ -110,22 +138,25 @@ def train_func(config):
...
     criterion = nn.CrossEntropyLoss()
 
-    results = []
-    for _ in range(epochs):
+    while True:
         train_epoch(train_loader, model, criterion, optimizer)
         result = validate_epoch(validation_loader, model, criterion)
-        session.report(result)
-        results.append(result)
+        checkpoint = Checkpoint.from_dict(
```
Comment (Yard1) [inline python/ray/train/examples/tune_cifar_torch_pbt_example.py:144]:
```suggestion
        checkpoint = TorchCheckpoint.from_dict(
```

**Review by Yard1 — COMMENTED** [review]
Let's make sure to use framework-specific checkpoints where applicable

File: python/ray/train/examples/tune_cifar_torch_pbt_example.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/examples/tune_cifar_torch_pbt_example.py:144]: I used a generic `Checkpoint.from_dict` rather than `TorchCheckpoint.from_model()` or `TorchCheckpoint.from_state_dict()` because there's no way to save the momentum optimizer state when going through the framework specific checkpoints.

Also, doing `TorchCheckpoint.from_dict` is a bit weird because I'm not setting the model state dict to the correct key, so I wouldn't be able to do `checkpoint.get_model()` later.

File: python/ray/train/examples/tune_cifar_torch_pbt_example.py [hunk 2]
```diff
@@ -88,11 +113,12 @@ def train_func(config):
     )
 
     with FileLock(".ray.lock"):
+        data_dir = config.get("data_dir", "~/data")
```
Comment (Yard1) [inline python/ray/train/examples/tune_cifar_torch_pbt_example.py:116]: can we use something like `os.path.expanduser("~/.ray.lock")` instead? Ideally tie it to the `data_dir`. If each worker runs in a separate dir, then they will not use the same lock files.

File: python/ray/train/examples/tune_cifar_torch_pbt_example.py [hunk 3]
```diff
@@ -110,22 +136,25 @@ def train_func(config):
     # Create loss.
     criterion = nn.CrossEntropyLoss()
 
-    results = []
-    for _ in range(epochs):
+    while True:
```
Comment (Yard1) [inline python/ray/train/examples/tune_cifar_torch_pbt_example.py:139]: what's the reason for using while here?

File: doc/source/tune/examples/includes/pbt_function.rst [hunk 4]
```diff
@@ -3,4 +3,9 @@
 PBT Function Example
 ~~~~~~~~~~~~~~~~~~~~
 
+The following script produces the following results. For a population of 8 trials,
+the PBT learning rate schedule roughly matches the optimal learning rate schedule.
+
```
Comment (xwjiang2010) [inline doc/source/tune/examples/includes/pbt_function.rst:8]: hmmm, how should I interpret this result?

File: python/ray/train/examples/tune_cifar_torch_pbt_example.py [hunk 5]
```diff
@@ -58,17 +58,42 @@ def validate_epoch(dataloader, model, loss_fn):
...
+        model_state = checkpoint_dict["model_state_dict"]
+        model.load_state_dict(model_state)
+
+        # Load in optimizer
+        optimizer_state = checkpoint_dict["optimizer_state_dict"]
+        optimizer.load_state_dict(optimizer_state)
+
+        # Optimizer configs (`lr`, `momentum`) are being mutated by PBT and passed in
+        # through config, so we need to update the optimizer loaded from the checkpoint
+        update_optimizer_config(optimizer, optimizer_config)
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/train/examples/tune_cifar_torch_pbt_example.py:92]: I think we just need to also save and load the learning rate scheduler state (which holds the epoch information), along with the optimizer. Then, PBT will perturb the LR, but it will still follow the same schedule. We could show how to do it in this example or another example maybe?

File: python/ray/train/examples/tune_cifar_torch_pbt_example.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/train/examples/tune_cifar_torch_pbt_example.py:139]: Yeah, I can change this back.

**Review by Yard1 — APPROVED** [review]
Thanks!

**Review by xwjiang2010 — APPROVED** [review]
Thanks, this is a great improvement!

**Review by richardliaw — APPROVED** [review]
